[Dorado] New Dorado Basecalling Workflow Terra

fraser-combe opened the pull request:
This PR closes #713 

🗑️ This dev branch should be deleted after merging to main.

## :brain: Summary
A new **Dorado Basecalling Workflow**, a GPU-accelerated pipeline for basecalling Oxford Nanopore `POD5` files. The workflow includes optional automatic model selection, SAM-to-BAM conversion, and demultiplexing into unique barcode fastq files, with outputs uploaded to a new user defined Terra table for further downstream analysis.


## :zap: Impacted Workflows/Tasks
This is a new workflow that does not impact any other workflows

This PR may lead to different results in pre-existing outputs: No

This PR uses an element that could cause duplicate runs to have different results: No

## :hammer_and_wrench: Changes
This PR introduces the following changes:
- **New Workflow**: Dorado Basecalling Workflow version 1.0.
- **Optional Inputs**: Added `use_auto_model` flag for automatic model selection.
- **Manual and Auto Model Options**: Supports both predefined models and automatic selection (`sup`, `hac`, `fast`).
- **SAM-to-BAM Conversion**: Integrated SAMTools task for efficient data handling.
- **Demultiplexing**: Added demux step to create barcode-specific FASTQ files.
- **Terra Integration**: Outputs transferred to Terra, with a table generated for downstream workflows.


### :gear: Algorithm
- **New Tasks**:
  1. **Dorado Basecall**: Converts `POD5` files to SAM using GPU acceleration. Uses a new Dorado Staph-B Docker image v0.80
  https://github.com/StaPH-B/docker-builds/tree/master/dorado/0.8.0
  2. **SAMTools Convert**: Converts SAM files to BAM.
  3. **Dorado Demultiplexing**: Creates barcode-specific FASTQ files.
  4. **File Transfer**: Uploads FASTQ files to Terra.
  5. **Terra Table Creation**: Generates Terra table from the uploaded FASTQ files.


### ➡️ Inputs
- **New Inputs**:
  - **use_auto_model** (Boolean): Enables automatic model selection.
  - **model_accuracy** (String): Specifies model accuracy if using auto-selection (`sup`, `hac`, `fast`).
  - **fastq_file_name** (String): Prefix for output FASTQ files.
  - **fastq_upload_path** (String): Path to Terra for uploading FASTQ files.
  - **kit_name** (String): Specifies sequencing kit for adapter/barcode trimming.

### ⬅️ Outputs
- **New Outputs**:
  - **basecalled_fastqs**: Array of FASTQ files generated from basecalling.
  - **demuxed_fastqs**: Array of FASTQ files generated from demultiplexing.
  - **logs**: Logs generated during the demux step.
  - **terra_table_tsv**: TSV file for uploading to Terra.

## :test_tube: Testing
- Tested locally with simulated `POD5` inputs and GPU resources.
- Tested in Terra with multiple pod5 files - produced expected barcodes
Test 1. With 9 Rabies pod5 files from 2 barcodes (manual model)
-https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_FCombe_sandbox/job_history/889322c2-19f0-4092-ac7f-4863e676b28a
![image](https://github.com/user-attachments/assets/9a231a2a-491b-4cf8-a7d2-736be6a33b40)

Test 2. 24 pod5 files from 2 barcodes (manual model)
https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_FCombe_sandbox/job_history/9bef28ea-82ba-4406-8545-f32de7e07e02

![image](https://github.com/user-attachments/assets/d74d545e-5789-4a1b-9623-ae488467f288)


test 3. 24 files from 2 barcodes (auto mode)
https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_FCombe_sandbox/job_history/cead789e-c737-4541-a6ed-d9b907493ee1

output terra table example
![image](https://github.com/user-attachments/assets/4d2bdfdd-87db-47b5-9798-ba7e8f299164)


- **Edge Case Handling**: Verified workflow behavior with missing inputs and unsupported models.
- **Terra Integration**: Confirmed successful transfer and Terra table generation with sample data.



### Suggested Scenarios for Reviewer to Test
1. **Basecalling with Auto Model Selection**: Run with the `use_auto_model` flag enabled.
2. **Manual Model Input**: Test with a specific `dorado_model` path and confirm outputs.
3. **Demultiplexing**: Verify barcode-specific FASTQ outputs.
4. **Edge Case**: Provide incomplete inputs (e.g., missing `kit_name`) to confirm error handling.
5. **Terra Table Generation**: Confirm Terra table creation and FASTQ uploads with valid inputs.

## :microscope: Final Developer Checklist
<!-- Please mark boxes [X] -->
- [X] The workflow/task has been tested and results, including file contents, are as anticipated
- [x] The CI/CD has been adjusted and tests are passing (Theiagen developers)
- [x] Code changes follow the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-51b66a47dde54c798f35d673fff80249)
- [x] Documentation and/or workflow diagrams have been updated if applicable (Theiagen developers only)

## 🎯 Reviewer Checklist
<!--  Indicate NA when not applicable  -->
- [ ] All changed results have been confirmed
- [ ] You have tested the PR appropriately (see the [testing guide](https://theiagen.notion.site/PR-Testing-Guide-Determining-Appropriate-Levels-of-Testing-4764e98a6aeb460185039c0896714590) for more information)
- [ ] All code adheres to the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-51b66a47dde54c798f35d673fff80249)
- [ ] MD5 sums have been updated
- [ ] The PR author has addressed all comments
- [ ] The documentation has been updated

fraser-combe commented:
Introduced scatter approach to basecalling step,
Documentation updated inputs and outputs and information boxes for key inputs such as sample naming convention in relation to terra table upload

New testing data 2 barcodes 47 pod 5 files successful
https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/de882ab4-21e7-465d-b70a-62f7456af509

AndrewLangvt commented on workflows/utilities/wf_dorado_basecalling.wdl:
```diff
@@ -0,0 +1,78 @@
+  }
+
+  output {
+    Array[File] fastq_files = dorado_demux.fastq_files
```
I'm concerned about the usability of this for labs. Wouldn't it be better to output this in such a way that it is accessible for follow-on workflows to use the fastq files (which generally take single inputs for read data, not arrays)?

fraser-combe replied:
We could adjust the output to use flatten, allowing individual FASTQ files to be directly accessible. Just checking, do you know if Terra workflows handle flatten consistently for outputting individual files rather than arrays? Here’s the current output configuration:
output {
    Array[File] individual_fastq_files = flatten(dorado_demux.fastq_files)
    File? terra_table_tsv = create_terra_table.terra_table_to_upload
}

AndrewLangvt replied:
You can disregard this comment. I overlooked the fact that you were also making these available via create_terra_table. Thanks @sage-wright for highlighting that to me. All good to leave them as an array.

AndrewLangvt commented on tasks/utilities/file_handling/task_transfer_files.wdl:
```diff
@@ -6,20 +6,20 @@ task transfer_files {
+    file_path_array="~{sep=' ' files_to_transfer}" 
 
-    gsutil -m cp -n ${file_path_array[@]} ~{target_bucket}
+    gcloud storage cp -n ${file_path_array[@]} ~{target_bucket}
```
gcloud is more efficient with data transfer, I believe. Is this the only reason to change to gcloud, or is there another use case?

fraser-combe replied:
This was an previous task that I use in the workflow but I can update this task as part of this PR as well

AndrewLangvt replied:
If there's an argument to change from gsutil to gcloud, I'm happy to discuss. Given the task_transfer_files is used by quite a few workflows, it would be better to avoid touching it unless we need to to minimize scope of potential impact.

sage-wright replied:
I'm for switching to gcloud storage. It's more efficient, and gsutil is a deprecated tool that Google no longer recommends

AndrewLangvt replied:
Wow, I need to get with the times. Deprecated, it is; let's move to gcloud!

AndrewLangvt commented on tasks/utilities/file_handling/task_transfer_files.wdl:
```diff
@@ -6,20 +6,20 @@ task transfer_files {
     Int cpu = 4
     Int memory = 8
-    String docker_image = "us-docker.pkg.dev/general-theiagen/theiagen/utility:1.1"
+    String docker_image = "gcr.io/google.com/cloudsdktool/google-cloud-cli:427.0.0-alpine"
```
do we have any internally hosted containers with gcloud that you could use, in lieu of an external one?

kapsakcj replied:
another option is to copy the gcr.io docker image over to our GAR and use that copy

the theiagen/utility:1.1 docker image is pretty large compared to the gcr.io docker image, I'm guessing that's might be part of why Fraser chose that one?

fraser-combe replied:
Im going to leave the transfer_files.wdl since this is used by other workflows. I assume it would be the preference to do this as a seperate PR for testing additional workflows affected

kapsakcj replied:
asking for clarity - @fraser-combe do you plan to undo the changes made to this WDL file?

It's currently conflicting with `main` due to the addition of `set -euo pipefail` in PR #666

kapsakcj replied:
we instead decided to go with a copy of the google.com docker image that's on our GAR. commit incoming shortly to update the default docker image for this task

fraser-combe commented on workflows/utilities/wf_dorado_basecalling.wdl:
```diff
@@ -0,0 +1,78 @@
+    String terra_project
+    String terra_workspace
+    String fastq_file_name
+    Int cpu = 8                      
```
Its ok, I forgot to remove these from the wf file so ill update this

kapsakcj commented on tasks/basecalling/task_dorado_demux.wdl:
```diff
@@ -0,0 +1,95 @@
+  }
+
+  command <<< 
+    set -e
```
might be good to adjust this to `set -euo pipefail` as there's one pipe used below

kapsakcj commented on tasks/basecalling/task_samtools_convert.wdl:
```diff
@@ -0,0 +1,42 @@
+    docker: docker
+    cpu: cpu
+    memory: "~{memory} GB"
+    disks: "local-disk ~{disk_size} HDD"
```
change `HDD` to `SSD`

fraser-combe commented on tasks/basecalling/task_dorado_basecall.wdl:
```diff
@@ -0,0 +1,53 @@
+    disk: disk_size + " GB"
+    gpuCount: 1
+    gpuType: "nvidia-tesla-t4"  }
+}
```
Yes I agree I originally had 1 retry but then removed it for testing so ill add back in

kapsakcj commented on tasks/basecalling/task_dorado_demux.wdl:
```diff
@@ -0,0 +1,95 @@
+    echo "### Zipping all FASTQ files ###"
+    for fastq in ~{fastq_file_name}-*.fastq; do
+      echo "Zipping $fastq"
+      gzip -f "$fastq"
```
if only `pigz` was available in the docker image 😢 this would go much faster. We can add in `pigz` for the next version of dorado added to StaPH-B's docker images

kapsakcj commented on tasks/basecalling/task_dorado_demux.wdl:
```diff
@@ -0,0 +1,95 @@
+    cpu: cpu
+    memory: "~{memory} GB"
+    disks: "local-disk ~{disk_size} SSD"
+  }
```
Similar thoughts here too. I would add `preemptible: 0` to ensure you do NOT get a preemptible VM for this computationally intensive task.

It may be worth also adding a line for `maxRetries: 1` to at least allow for retrying once more with additional memory in case of a memory failure. We usually do 3 maxRetries but that could get expensive quickly since you set default memory to 16

kapsakcj commented on tasks/basecalling/task_dorado_demux.wdl:
```diff
@@ -0,0 +1,95 @@
+    String fastq_file_name
+    Int cpu = 4 
+    Int memory = 16
+    Int disk_size = 50
```
storage is cheap and there's a good chance of running out of disk space if input BAMs and output FASTQ files are large. I'd increase default disk_size to at least 100 GB. Imagine a full GridION run with 5 flowcells worth of data being basecalled in a single workflow. Could easily surpass 50GB

kapsakcj commented on tasks/basecalling/task_samtools_convert.wdl:
```diff
@@ -0,0 +1,42 @@
+      bam_file="${bam_output}/${base_name}.bam"
+
+      echo "Converting SAM to BAM: $sam_file -> $bam_file"
+      samtools view -bS "$sam_file" > "$bam_file" || { echo "ERROR: samtools failed to convert $sam_file to BAM" >&2; exit 1; }
```
not sure if samtools automatically uses all cpus available so I would add the option to pass in number of cpus: `samtools view -@ $(nproc) ....`

kapsakcj commented on tasks/basecalling/task_dorado_basecall.wdl:
```diff
@@ -0,0 +1,53 @@
+      "~{input_file}" \
+      --kit-name ~{kit_name} \
+      --emit-sam \
+      --no-trim \
```
Is there any reason why we are defaulting to `--no-trim`?

Might be good to set this as a string input in case we want to enable trimming in the future?

in the inputs:
`String? trim_mode = "none" # options are: "none", "all", "adapters", "primers"`

and in the command block:
`--trim ~{trim_mode} \`

Here's the relevant part of the `dorado basecaller --help` output:
```
Trimming arguments (detailed usage):
  --no-trim                   Skip trimming of barcodes, adapters, and primers. If option is not chosen, trimming of all three is enabled. 
  --trim                      Specify what to trim. Options are 'none', 'all', 'adapters', and 'primers'. Default behaviour is to trim all detected adapters, primers, or barcodes. Choose 'adapters' to just trim adapters. The 'primers' choice will trim adapters and primers, but not barcodes. The 'none' choice is equivelent to using --no-trim. Note that this only applies to DNA. RNA adapters are always trimmed. [nargs=0..1] [default: ""]
```

fraser-combe replied:
It needs to be no trim so the barcodes are maintained for the demultiplexing task

kapsakcj commented on mkdocs.yml:
```diff
@@ -52,6 +52,7 @@ nav:
                - Zip_Column_Content: workflows/data_export/zip_column_content.md
             - Standalone:
                - Cauris_CladeTyper: workflows/standalone/cauris_cladetyper.md
+               - Dorado_Bsecalling: workflows/standalone/dorado_basecalling.md
```
typo here

fraser-combe commented:
Ready for review, thank you!

kapsakcj commented on docs/workflows/standalone/dorado_basecalling.md:
```diff
@@ -0,0 +1,111 @@
+| **Terra Task Name** | **Variable** | **Type** | **Description** | **Default Value** | **Terra Status** |
+|---|---|---|---|---|---|
+| dorado_basecalling_workflow | **input_files** | Array[File] | Array of `POD5` files for basecalling | None | Required |
+| dorado_basecalling_workflow | **dorado_model** | String | Model accuracy or full model name (default: 'sup') | "sup" | Optional |
```
I think this line is slightly inaccurate? `sup`, `hac`, and `fast` are **model speeds** , rather than the actual model itself: https://github.com/nanoporetech/dorado?tab=readme-ov-file#model-selection-foreword

kapsakcj replied:
not a big deal, but it might be good to stay consistent with ONT's language

kapsakcj commented on docs/workflows/standalone/dorado_basecalling.md:
```diff
@@ -0,0 +1,111 @@
+| dorado_basecalling_workflow | **terra_workspace** | String | Terra workspace for final FASTQ file upload | None | Required |
+| dorado_basecalling_workflow | **paired_end** | Boolean | Indicates if data is paired-end | false | Optional |
+| dorado_basecalling_workflow | **assembly_data** | Boolean | Indicates if the data is for assembly | false | Optional |
+| dorado_basecalling_workflow | **file_ending** | String? | File extension pattern for identifying files (e.g., ".fastq.gz") | None | Optional |
```
Would it make sense to re-order the rows in this table so that all required inputs are at the top, followed by optional inputs below (with them being in alphabetical order, as Terra displays them)?

Not sure if this is done for other workflow inputs tables but might be slightly helpful to the user

kapsakcj commented on tasks/basecalling/task_dorado_basecall.wdl:
```diff
@@ -0,0 +1,54 @@
+      --no-trim \
+      --output-dir "$sam_output" \
+      --verbose || { echo "ERROR: Dorado basecaller failed for ~{input_file}"; exit 1; }
+
```
consider re-naming output SAM files to avoid duplicate SAM filenames produced across different shards

kapsakcj replied:
perhaps `basename ~{input_file}` so final resulting file is the `basename.sam`

sage-wright replied:
I think he's doing this see line 23

sage-wright replied:
wait, i think that needs to be updated so line 35 references $sam_file instead

kapsakcj replied:
Fraser and I chatted about this and believe we have a solution. Dorado allows you to name the output directory, but not the specific output SAM file (AFAIK)

so I think he's going to add some code to rename the output SAM file based on the input POD5 filename which should be a unique string instead of `$sam_output/call_<datetime-stamp>.sam` which is where I found duplicate filenames because of the very slim chance that dorado ran at the exact same time on 2 separate input pod5 files :/

kapsakcj commented on tasks/basecalling/task_dorado_demux.wdl:
```diff
@@ -0,0 +1,101 @@
+
+      echo "Renaming $fastq_file to $final_fastq"
+
+      if [ -f "$final_fastq" ]; then
```
`-f` should work here, but if not try `-s` which means `True if FILE exists and has a size greater than zero.`

https://tldp.org/LDP/Bash-Beginners-Guide/html/sect_07_01.html

kapsakcj replied:
disregard, `-f` works fine here

kapsakcj commented on tasks/basecalling/task_dorado_demux.wdl:
```diff
@@ -0,0 +1,102 @@
+        --kit-name "$kit_name" \
+        --emit-fastq \
+        --emit-summary \
+        --verbose > "$demux_dir/demux_${base_name}.log" 2>&1 || {
```
just want to note that `dorado demux` does allow you to control threads/cpus used and that by default it will use all available threads. I don't forsee any issue with this, unless running on a shared compute system (i.e. not Terra). I don't think it's necessary to update the code for this now

```
-t, --threads          Combined number of threads for barcoding and output generation. Default uses all available threads. [nargs=0..1] [default: 0]
```

kapsakcj commented on tasks/basecalling/task_dorado_demux.wdl:
```diff
@@ -0,0 +1,102 @@
+
+        # Verify if the merged file already exists
+        if [ -f "$final_fastq" ]; then
+          echo "Appending to existing $final_fastq"
```
TY for adding in the echo statements in this section for logging. It will be helpful if we ever need to debug basecalling & demultiplexing

kapsakcj commented:
@fraser-combe Consider adding a few more of the typical outputs to the tasks (and thus the workflow too):
- `String dorado_docker` for basecall and demux tasks
- `String dorado_version` for basecall and demux tasks
- `String dorado_model` for the explicit model used in basecall task, for example: `dna_r10.4.1_e8.2_400bps_sup@v5.0.0` would be useful to the user to know exactly which model was used
  - one option for capturing this string is to `| tee dorado-basecall.log` the STDOUT of `dorado basecaller` to a log file (unless dorado already produces a log file) and parse it from there.
- `Array[File] dorado_demux_summary_tsvs` for the demux task - is this file useful for debugging? I would recommend to keep as a task output, but not a workflow output.
- similar outputs for the `samtools_convert` task: `samtools_docker`, `samtools_version`
- workflow level outputs to add (to mirror what we do with other TheiaProk/TheiaCov wfs):
  - `String dorado_basecalling_phb_version`
  - `String dorado_basecalling_phb_analysis_date`

^ which begs the question - should we add in the typical `version_capture` task that we use for theiacov/theiaprok wfs? I think we should as it's a relatively easy add.

All of this will require documentation updates as well. 

I'm happy to help with all of the above if you're OK with me pushing commits @AndrewLangvt and @fraser-combe 

----

FYI After your recent commits I ran [this wf on a test dataset and had great success](https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/4ad1ab46-1d02-4fa2-8dc6-a641114a3ad1) ✅  with recovering all the FASTQs that I was expecting. I was able to assemble some genomes with [TheiaProk_ONT so that's a good test to show things are running as expected](https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/601f5cfb-83e9-460d-bd96-9919bc93a95d) ✅ 🎉 ! Very nice work with this workflow so far. It's great, but we can always make it a little bit better

sage-wright commented on docs/workflows_overview/workflows_alphabetically.md:
```diff
@@ -16,6 +16,7 @@ title: Alphabetical Workflows
 | [**Core_Gene_SNP**](../workflows/phylogenetic_construction/core_gene_snp.md) | Pangenome analysis | Bacteria | Set-level | Some optional features incompatible, Yes | v2.1.0 | [Core_Gene_SNP_PHB](https://dockstore.org/workflows/github.com/theiagen/public_health_bioinformatics/Core_Gene_SNP_PHB:main?tab=info) |
 | [**Create_Terra_Table**](../workflows/data_import/create_terra_table.md)| Upload data to Terra and then run this workflow to have the table automatically created | Any taxa | | Yes | v2.2.0 | [Create_Terra_Table_PHB](https://dockstore.org/workflows/github.com/theiagen/public_health_bioinformatics/Create_Terra_Table_PHB:main?tab=info) |
 | [**CZGenEpi_Prep**](../workflows/phylogenetic_construction/czgenepi_prep.md)| Prepare metadata and fasta files for easy upload to the CZ GEN EPI platform. | Monkeypox virus, SARS-CoV-2, Viral | Set-level | No | v1.3.0 | [CZGenEpi_Prep_PHB](https://dockstore.org/workflows/github.com/theiagen/public_health_bioinformatics/CZGenEpi_Prep_PHB:main?tab=info) |
+| [**Dorado_Basecalling**](../workflows/standalone/dorado_basecalling.md)| GPU-accelerated basecalling of Oxford Nanopore sequencing data | Any taxa | Sample-level | Yes | v1.0.0 | [Dorado_Basecalling_PHB](https://dockstore.org/workflows/github.com/theiagen/public_health_bioinformatics/Dorado_Basecalling_PHB:main?tab=info) |
```
can you update the version updated number here to be 2.3.0
and in the other files too

sage-wright commented on docs/workflows/standalone/dorado_basecalling.md:
```diff
@@ -0,0 +1,111 @@
+
+| **Workflow Type** | **Applicable Kingdom** | **Last Known Changes** | **Command-line Compatibility** | **Workflow Level** |
+|---|---|---|---|---|
+| [Standalone](../../workflows_overview/workflows_type.md/#standalone) | [Any Taxa](../../workflows_overview/workflows_kingdom.md/#any-taxa) | v2.2.1 | Yes | Sample-level |
```
can you update the last known changes to v2.3.0

fraser-combe commented:
outputs working, see ,https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/79417a5e-da8c-4fdc-aa61-f28de8490bba 
![image](https://github.com/user-attachments/assets/085ef669-7d3f-4415-bf34-87ae57ac0e80)

kapsakcj commented on tasks/basecalling/task_dorado_basecall.wdl:
```diff
@@ -0,0 +1,80 @@
+
+    # Run Dorado basecaller and log output
+    dorado basecaller \
+      "~{dorado_model}" \
```
I tried to override the `dorado_model` input parameter with the explicit fast model name `dna_r10.4.1_e8.2_400bps_fast@v5.0.0` and I got an error saying that the model file was not found:
```
[2024-11-16 15:47:11.731] [info] Running: "basecaller" "dna_r10.4.1_e8.2_400bps_fast@v5.0.0" "/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/FAW95599_9a7a72fa_fad33a0d_0.pod5" "--kit-name" "SQK-RBK114-96" "--emit-sam" "--no-trim" "--output-dir" "output/sam_FAW95599_9a7a72fa_fad33a0d_0/" "--verbose"
[2024-11-16 15:47:11.743] [error] Model path does not exist at: 'dna_r10.4.1_e8.2_400bps_fast@v5.0.0' - Please download the model or use a model complex
[2024-11-16 15:47:11.743] [error] finalise() not called on a HtsFile.
```

since all the dorado model files are located at `/dorado_models` within the docker image filesystem this may need to be accounted for in the code. When I prefixed it and provided `/dorado_models/dna_r10.4.1_e8.2_400bps_fast@v5.0.0` it ran successfully.

So I'm thinking we need something like this in the code:
```
if [ "~{dorado_model}" == "fast" || "~{dorado_model}" == "hac" || "~{dorado_model}" == "sup" ]; then
  dorado_model=~{dorado_model}
  echo "DEBUG: dorado_model is set to: $dorado_model"
else
  dorado_model="/dorado_models/~{dorado_model}"
  echo "DEBUG: dorado_model is set to: $dorado_model"
fi
```
and then in the dorado command, need to swap `dorado basecaller ~{dorado_model}` with the bash variable `dorado basecaller $dorado_model` so that it handles both cases

kapsakcj commented:
I will test 3 different workflows and report back:

1. using `fast` as dorado_model input string
2. using `dna_r10.4.1_e8.2_400bps_fast@v5.0.0`
3. using `sup` (as this will be the recommended input param for our users)

kapsakcj commented on tasks/basecalling/task_dorado_demux.wdl:
```diff
@@ -0,0 +1,116 @@
+    Array[File] fastq_files = glob("merged_output/*.fastq.gz")
+    String dorado_docker = docker
+    String dorado_version = read_string("DORADO_VERSION")
+    File dorado_demux_log = "dorado_demux.log" 
```
I could be wrong here, but I think `dorado_demux.log` is the same filename as the log file that Cromwell automatically creates based off the task name `dorado_demux`

When I looked at this output file in the execution directory, it appears that it's the typical Cromwell log file, and not the log file created as part of this task on line 28.

I would consider renaming this file to a unique filename if you want to capture it separate from the Cromwell-produced log file.

kapsakcj commented:
TheiaProk_ONT ran successfully on the FASTQs produced by my test above with SUP dorado model 👍 https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/238d0f1f-fe13-4823-8846-b0774fb75e0c

More confirmation the FASTQs produced by this wf are valid for downstream processing

kapsakcj commented on docs/workflows/standalone/dorado_basecalling.md:
```diff
@@ -0,0 +1,114 @@
+
+### Workflow Tasks
+
+This workflow is composed of several tasks to process, basecall, and analyze rabies genome data:
```
remove mention of rabies

fraser-combe commented:
Added new task that allows user to upload pod5 files to the Data Uploader in Terra and provide the link to the Google bucket as an import to the workflow.

The new task will place combine the file paths into an array and pass to basecalling task.

Documentation has been updated with visuals for user to use the data uploaded and copying bucket link to workflow input

Tested with small number of pod5 files
https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/fbb56bc8-7d91-4f82-a246-40babd601303

Tested with large number of Pod5 files
https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/205e0035-a506-452a-841a-902eaba6e9c8

kapsakcj commented on tasks/utilities/file_handling/task_transfer_pod5_files.wdl:
```diff
@@ -0,0 +1,34 @@
+    String pod5_bucket_path             # GCS bucket path containing `.pod5` files (e.g., "gs://your-terra-bucket/pod5_uploads/")
+    Int disk_size = 100
+    Int memory = 32
+    Int cpu = 8
```
Recommend decreasing the default cpus and memory for this task since it's not actually copy/transferring files anywhere, but rather just listing them from the input bucket path

Perhaps use 1 cpu as default and 4 GB memory.

Also suggest changing the task name on line 3 as well as the WDL filename, it's a bit misleading. (also means updating the import statement in the workflow WDL file.

last suggestion is to enable premptible VMs as the runtime is likely very short for this task. Even if the VM were pre-empted, it shouldn't take long to run. Your test with a large number of POD5 files only took 2 min to run.

kapsakcj commented:
I appreciate your fortitude in making even further changes to this workflow. These changes will seriously simplify the setup process for the end user and save everyone lots of time.

The doc updates look great, the screenshots and section on uploading POD5 files and getting the input GSURI for where the files were uploaded look great. Straightforward and easy to understand & follow (in my opinion).

I'm launching a test here in Terra, but assuming it won't finish before I go on PTO for the holidays. Given your recent tests & my previous tests, I'm pretty confident it will run successfully. Please check the logs & outputs in my absence  https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/1c0618d9-560c-4741-a816-2c8bd7b89f15

- The new task for listing POD5 files ran successfully and the array was successfully passed into the dorado basecalling scatter block of the wf 👍 

Don't wait for me if the you/dev team wants to merge this PR before I'm back in office.

kapsakcj commented:
>Please check the logs & outputs in my absence https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/1c0618d9-560c-4741-a816-2c8bd7b89f15

OK, follow up to the above workflow ^

WF ran successfully ✅. It produced valid FASTQs, created a Terra data table with links to these FASTQs, and successfully uploaded them to Terra.

I ran the FASTQs through TheiaProk_ONT and outputs were generated successfully, showing that the dorado output FASTQ files were valid and useable for downstream analysis

kapsakcj commented on docs/workflows/standalone/dorado_basecalling.md:
```diff
@@ -0,0 +1,136 @@
+    Lists `.pod5` files in the specified GCS bucket and passes their paths to the basecalling task.
+
+??? task "`Dorado Basecalling`: Converts `POD5` files to 'SAM' files"
+    The basecalling task takes `POD5` files as input and converts them into 'SAM' format using the specified model. This step leverages GPU acceleration for efficient processing.
```
Might be good to mention that barcode trimming is purposefully disabled in the basecalling step so that the demultiplexing step later in the wf can accurately demultiplex (and trim barcodes)

kapsakcj commented on docs/workflows/standalone/dorado_basecalling.md:
```diff
@@ -0,0 +1,136 @@
+    Once the SAM files are generated, this task converts them into BAM format, optimizing them for downstream applications and saving storage space.
+
+??? task "`Dorado Demultiplexing`: Produces barcode-specific FASTQ files"
+    This task demultiplexes the BAM files based on barcodes, generating individual FASTQ files for each barcode to support further analyses.
```
I recommend mentioning that this step does trim barcodes by default.

If we do decide to enable the option to disable barcode trimming, we could also indicate that here too. Could say something like: 

"Barcode trimming is enabled by default, but could be disabled by setting the optional input variable `disable_barcode_trimming` to `true`"

fraser-combe commented:
Curtis thanks for testing this with users and outlining these updates, it helps improve the workflow

- Included recommended updates from users related to -no-trim option for dorado demultiplexing for potential to inspect reads set false for default 
- Included the option for users to provide a custom barcode FASTA file for trimming with their custom primers instead of Dorado default
- Updated Documentation to reflect the no trimming option and custom primer optional file

Testing is ongoing and I will update comment once complete

AndrewLangvt commented:
Good catch on some of the updates you recently made to the no trim option & also custom primers @fraser-combe. Structurally, I think this is nearly ready. @kapsakcj do you have bandwidth to do some additional testing of these new features?

fraser-combe commented:
Thanks, planning to do a final check with custom primers today as we now have the data from users. So once that is successful Ill link to the submissions for Curtis to review

kapsakcj commented:
@AndrewLangvt @fraser-combe Let me check with team leads first, but given the green light I should have time later this week or early next week to review Fraser's tests and do some of my own testing

fraser-combe commented:
Ready for review, 
Added a new task dorado_trim.wdl that takes the custom primers fasta file and trims any sequences that are identified after the dorado_demultiplexing step (optional file if not present skips) tested [here ](https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/4a797849-c8eb-453b-95b4-40a512d21f0d )and confirmed locally
No trim functionality tested and successful locally and in terra  
[no trim set to false](https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/f40101df-40cb-49c6-a9d9-4ba6d2ae65fd)
[no trim set to true](https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/7c92b7ca-1369-4891-88de-dbbfdee0f58d)

Updated Dorado to v 0.9.0 [tested here](https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/4ba4d59c-6057-4062-abf4-b6336f437b3c/8c81c946-c7a2-48b3-b472-eef882ef9c84) (currently still running)

sage-wright commented:
Hey Fraser! I haven't had any chance to look at this PR yet, so here are some of my thoughts. Some of these things **do not** need to be implemented, but could be "nice-to-have" and so I've indicated them as such and removed the check box since they're mainly ideas for the future

1. Documentation suggestions:
  - [ ] fix the typo in `dorado_input_seelction.png` file name
  - [ ] update the "last known changes" for dorado in the workflows_overviews and quick facts section to be vX.X.X
  - [ ] in workflows_type.md, cauris_cladetyper's last known changes has been accidentally overwritten to be v1.0.0 when it should be vX.X.X. 
  - [ ] looks like a similar thing happened in workflows_kingdom.md, since kraken2's entry is accidentally duplicated. I think the version directly beneath the entry for Dorado_Basecalling should be removed
  The rest of these apply to `dorado_basecalling.md`:
  - _NICE TO HAVE; NOT NECESSARY_: please clarify the language in the introduction since the workflow doesn't utilize the terra data uploader itself. The user uploads the files outside of the workflow and the workflow scrapes the provided path to find the data
  - _NICE TO HAVE; NOT NECESSARY_: consider making "Example Manual Models" a level 4 header instead of a level 3 header so that it is beneath the "Model Type Selection" header
  - [ ] Turn the toggle for the supported kit names into a heading, or alternatively, expand the toggle by default. It’s easy to accidentally scroll past this.
  - [ ] move the "uploading pod5 files" section to be the very first section after the introduction and remove the `"` from the end of that header, 
     - _NICE TO HAVE; NOT NECESSARY_: I would capitalize POD5 in the header as well to make sure the capitalization is the same everywhere
  - [ ] move the "Detailed Input Information" box and the "File Naming Guidelines" to be directly underneath the Inputs heading since this information is really useful and should be higher up!
      - [ ] in the detailed input information, the fastq_file name example (`"project001-"`)should have the `-` removed since that (the `-`) is automatically added in the code
      - [ ] in the file naming guidelines, please update the examples since based on the described behavior, if you provide the accepted prefix in the example, the output file names would be "projectname-barcode01.fastq.gz-barcodeXX.fastq.gz"
  - [ ] In the inputs table, please update the runtime parameters to use the language described in the style guide for the docs
  - [ ] In the workflow tasks section, please add the expected `!!! techdetails` tables to these tasks as well
  
2. dorado_basecall suggestions:
  - [ ] The output of this task is natively a BAM file, so instead of putting `--emit-sam` you can remove that parameter _and_ skip the convert_sam_to_bam task! Efficiency!! Make sure to update all of the `sam` references to `bam` accordingly
  - make sure to include all the runtime parameters to match the other tasks (the `disk: disksize + "GB"` ones or whatever they are) (and maybe do this for the other tasks if applicable)
  
3. dorado_demx.wdl suggestions:
  - [ ] implement curtis' suggestion to add the `--threads` input parameter! this will speed up runtime
  - _NICE TO HAVE; NOT NECESSARY_: maybe rename the `fastq_file_name` variable to be `output_file_prefix` to be more clear about its purpose
  - [ ] remove the `date` command at the very end
  
5. list_pod5_files wdl suggestions:
  - this task looks great!! I can see us using this task in a ton of other scenarios, but we would need to generalize this. Could you please: 
      - [ ] create a `file_ending` variable to grep for instead of the hard-coded `.pod5` ending
      - [ ] change all pod5 references to be generic
      
6. wf_dorado_basecalling.wdl suggestions:
  - [ ] rename `notrim` to specifically reference the `demux` task (something like `demux_no_trim`) since we _are_ trimming in a separate task and this could be confusing
  - [ ] fix whitespaces to match style guide
  - [ ] remove the `paired_end` and `assembly_data` and `file_ending` input variables and hard-code the `create_terra_table` _task_ inputs to have `paired_end = false,` and `assembly_data = false,` (because those will always be false in this workflow). Also, hard-code the `file_ending` to be `".fastq.gz"` as that is the file ending you are giving to all of the generated fastqs and letting the user have the ability to modify that may result in task failure

I have a final comment that would result in an efficiency improvement, but it's a bigger lift and functionally would behave exactly the same as what you're already doing so I don't want you to implement it. But I thought I would mention it here for the sake of our posterity. 

Instead of doing a transfer_files -> create_terra-table, I think we could save a fair bit of runtime by sending the array of fastq files directly into a new task that will make a terra table using that array and then upload that table. It would act like `create_terra_table` but instead of finding data in a directory, it would take the data from an input array. 

This looks great!!

fraser-combe commented:
Ready for review, 
Added a new task dorado_trim.wdl that takes the custom primers fasta file and trims any sequences that are identified after the dorado_demultiplexing step (optional file if not present skips) tested [here ](https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/4a797849-c8eb-453b-95b4-40a512d21f0d )and confirmed locally
No trim functionality tested and successful locally and in terra  for dorado_demux update
[no trim set to false](https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/f40101df-40cb-49c6-a9d9-4ba6d2ae65fd)
[no trim set to true](https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/7c92b7ca-1369-4891-88de-dbbfdee0f58d)

Sage comments updated throughout - thanks some ideas I had not thought of

Documentation updated to reflect new inputs and new trim task
removed samtools convert task with updated dorado now outputting bams

Updated Dorado to v 0.9.0 and updated all docker images to current version in basecall, demux and trim steps [tested here](https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/898374f0-0a37-4c23-a257-c72d4301935c/6a7c5017-4d64-448e-b120-1ad1880232ed) (currently still running)

sage-wright commented on docs/workflows/standalone/dorado_basecalling.md:
```diff
@@ -0,0 +1,193 @@
+
+### Supported Kit Names:
+
+    Ensure you use an accepted kit name in the `kit_name` parameter. The following kit names are supported in Dorado:
```
Please unindent this section so that all of this header's content will not be read as a code block (see screenshot for how this looks on the serve preview)

![image](https://github.com/user-attachments/assets/9404437e-c442-4952-a307-79ab4ae7693d)

fraser-combe replied:
My last push on Friday for the docs didnt go up let me fix and re push

sage-wright commented on docs/workflows/standalone/dorado_basecalling.md:
```diff
@@ -0,0 +1,193 @@
+     !!! note "Barcode Trimming"
+        Barcode trimming is purposefully disabled during the basecalling step to ensure accurate demultiplexing in subsequent workflow steps.
+
+    !!! techdetails "Dorado Basecalling Technical Details"
```
Some of these techdetails blocks aren't rendering properly due to indentation. Can you fix those up please?

fraser-combe replied:
re pushed my commit thanks

AndrewLangvt commented on docs/workflows_overview/workflows_alphabetically.md:
```diff
@@ -19,6 +19,7 @@ title: Alphabetical Workflows
 | [**Core_Gene_SNP**](../workflows/phylogenetic_construction/core_gene_snp.md) | Pangenome analysis | Bacteria | Set-level | Some optional features incompatible, Yes | v2.1.0 | [Core_Gene_SNP_PHB](https://dockstore.org/workflows/github.com/theiagen/public_health_bioinformatics/Core_Gene_SNP_PHB:main?tab=info) |
 | [**Create_Terra_Table**](../workflows/data_import/create_terra_table.md)| Upload data to Terra and then run this workflow to have the table automatically created | Any taxa | | Yes | v2.2.0 | [Create_Terra_Table_PHB](https://dockstore.org/workflows/github.com/theiagen/public_health_bioinformatics/Create_Terra_Table_PHB:main?tab=info) |
 | [**CZGenEpi_Prep**](../workflows/phylogenetic_construction/czgenepi_prep.md)| Prepare metadata and fasta files for easy upload to the CZ GEN EPI platform. | Monkeypox virus, SARS-CoV-2, Viral | Set-level | No | v1.3.0 | [CZGenEpi_Prep_PHB](https://dockstore.org/workflows/github.com/theiagen/public_health_bioinformatics/CZGenEpi_Prep_PHB:main?tab=info) |
+| [**Dorado_Basecalling**](../workflows/standalone/dorado_basecalling.md)| GPU-accelerated basecalling of Oxford Nanopore sequencing data | Any taxa | Sample-level | Yes | v2.3.0 | [Dorado_Basecalling_PHB](https://dockstore.org/workflows/github.com/theiagen/public_health_bioinformatics/Dorado_Basecalling_PHB:main?tab=info) |
```
version should be vX.X.X as this will be a new workflow. We'll update that to the next release tag. If it's left as v2.3.0, we might miss making that change just before the next release.

AndrewLangvt commented on docs/workflows_overview/workflows_kingdom.md:
```diff
@@ -17,6 +17,7 @@ title: Workflows by Kingdom
 | [**Concatenate_Column_Content**](../workflows/data_export/concatenate_column_content.md) | Concatenate contents of a specified Terra data table column for many samples ("entities") | Any taxa | Set-level | Yes | v2.1.0 | [Concatenate_Column_Content_PHB](https://dockstore.org/workflows/github.com/theiagen/public_health_bioinformatics/Concatenate_Column_Content_PHB:main?tab=info) |
 | [**Concatenate_Illumina_Lanes**](../workflows/standalone/concatenate_illumina_lanes.md)| Concatenate Illumina lanes for a single sample | Any taxa | Sample-level | Yes | v2.3.0 | [Concatenate_Illumina_Lanes_PHB](https://dockstore.org/workflows/github.com/theiagen/public_health_bioinformatics/Concatenate_Illumina_Lanes_PHB:main?tab=info) |
 | [**Create_Terra_Table**](../workflows/data_import/create_terra_table.md)| Upload data to Terra and then run this workflow to have the table automatically created | Any taxa | | Yes | v2.2.0 | [Create_Terra_Table_PHB](https://dockstore.org/workflows/github.com/theiagen/public_health_bioinformatics/Create_Terra_Table_PHB:main?tab=info) |
+| [**Dorado_Basecalling**](../workflows/standalone/dorado_basecalling.md)| GPU-accelerated basecalling of Oxford Nanopore sequencing data | Any taxa | Sample-level | Yes | v2.3.0 | [Dorado_Basecalling_PHB](https://dockstore.org/workflows/github.com/theiagen/public_health_bioinformatics/Dorado_Basecalling_PHB:main?tab=info) | 
```
version should be vX.X.X as this will be a new workflow. We'll update that to the next release tag. If it's left as v2.3.0, we might miss making that change just before the next release.

AndrewLangvt commented on docs/workflows_overview/workflows_type.md:
```diff
@@ -98,6 +98,7 @@ title: Workflows by Type
 | **Name** | **Description** | **Applicable Kingdom** | **Workflow Level** | **Command-line Compatibility**[^1] | **Last Known Changes** | **Dockstore** |
 |---|---|---|---|---|---|---|
 | [**Cauris_CladeTyper**](../workflows/standalone/cauris_cladetyper.md)| C. auris clade assignment | Mycotics | Sample-level | Yes | vX.X.X | [Cauris_CladeTyper_PHB](https://dockstore.org/workflows/github.com/theiagen/public_health_bioinformatics/Cauris_CladeTyper_PHB:main?tab=info) |
+| [**Dorado_Basecalling**](../workflows/standalone/dorado_basecalling.md)| GPU-accelerated basecalling of Oxford Nanopore sequencing data | Any taxa | Sample-level | Yes | v2.3.0 | [Dorado_Basecalling_PHB](https://dockstore.org/workflows/github.com/theiagen/public_health_bioinformatics/Dorado_Basecalling_PHB:main?tab=info) |
```
version should be vX.X.X as this will be a new workflow. We'll update that to the next release tag. If it's left as v2.3.0, we might miss making that change just before the next release.

AndrewLangvt commented on docs/workflows/standalone/dorado_basecalling.md:
```diff
@@ -0,0 +1,189 @@
+| version_capture | **docker** | String | The Docker container to use for the task | "us-docker.pkg.dev/general-theiagen/theiagen/alpine-plus-bash:3.20.0" | Optional |
+| version_capture | **timezone** | String | Set the time zone to get an accurate date of analysis (uses UTC by default) | | Optional |
+
+!!! info "Detailed Input Information"
```
please move this and the "File Naming Guidelines" to be directly below the Inputs header, as this is very important to the user (and many users stop reading after they get to the inputs table).

AndrewLangvt commented on docs/workflows/standalone/dorado_basecalling.md:
```diff
@@ -0,0 +1,189 @@
+This workflow is composed of several tasks to process, basecall, and analyze Oxford Nanopore `POD5` files:
+
+??? task "`Transfer POD5 Files`: Transfers `POD5` files to Terra"
+    Lists `.pod5` files in the specified GCS bucket and passes their paths to the basecalling task.
```
can you add the link to the task, as you do in other `techdetails` sections?

AndrewLangvt commented on docs/workflows/standalone/dorado_basecalling.md:
```diff
@@ -0,0 +1,189 @@
+        | Software Documentation | <https://github.com/nanoporetech/dorado> |
+
+??? task "`FASTQ File Transfer`: Transfers files to Terra"
+    After demultiplexing, the FASTQ files are uploaded to the Terra workspace bucket for storage and potential use in other workflows.
```
can you add the link to the task, as you do in other `techdetails` sections? Same with the Terra Table Creation below this.

AndrewLangvt commented on tasks/basecalling/task_dorado_trim.wdl:
```diff
@@ -0,0 +1,41 @@
+    mkdir -p "$output_dir"
+
+    # Loop through each FASTQ file and apply Dorado trim
+    for fq in ~{sep=" " fastq_files}; do
```
refer to [style guide](https://www.notion.so/theiagen/Style-Guide-WDL-Workflow-Development-51b66a47dde54c798f35d673fff80249?pvs=4#8fe451ba440f44d6bea376e4d2df9071) for indentation spacing

AndrewLangvt commented on workflows/utilities/data_import/wf_create_terra_table.wdl:
```diff
@@ -6,10 +6,10 @@ workflow create_terra_table {
     String data_location_path
-    Boolean paired_end
-    Boolean assembly_data
+    Boolean paired_end = false
```
I appreciate that this was carry-over from my recommendation that you could just leave your defaults defined at the task level (still accurate). However, modifying this will now impact many other workflows. We should, in this case with create_terra_table, avoid adjusting the task defaults to ensure minimal unnecessary impact across PHB.

AndrewLangvt commented on tasks/utilities/file_handling/task_list_pod5_files.wdl:
```diff
@@ -0,0 +1,34 @@
+version 1.0
+
+task list_pod5_files {
```
A comment @sage-wright had here that I really liked was generalizing this task, such that it could be used by other workflows. I.e. pass it the extension one wants to search for, and the process would list files with the corresponding extension in the provided bucket. It definitely would be something we could use elsewhere, and probably wouldn't be overly painful to modify.

AndrewLangvt reviewed (changes requested):
A few follow ups that are outstanding from Sage's initial comments.

fraser-combe commented:
[Reconfirmed dorado successful ](https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/3186291f-6b36-470c-9749-582e01264075)

AndrewLangvt commented:
@kapsakcj are you able to take another look at this, following your previous comments & to ensure it will meet the needs of CDPH?

kapsakcj commented on tasks/basecalling/task_dorado_demux.wdl:
```diff
@@ -0,0 +1,105 @@
+
+    echo "### Compressing merged FASTQ files ###"
+    for final_fastq in merged_output/*.fastq; do
+      gzip -f "$final_fastq"
```
`pigz` is now available in the docker image `us-docker.pkg.dev/general-theiagen/staphb/dorado:0.9.0-cuda12.2.0`

I might recommend changing this line to `pigz "$final_fastq"` or perhaps remove the whole `for` loop and just do `pigz merged_output/*.fastq`

If you get rid of the for loop, `pigz` will essentially compress all files in-place and add the suffix `.gz`, so you'll end up with `merged_output/<filenames>.fastq.gz`
